LSPS2/README.md

```diff
@@ -0,0 +1,1001 @@
+channel open.
+
+Once any PPPs are used to pay for a new channel open, the
+hold-incoming state is cleared and normal operation resumes
+(i.e. any incoming payments are forwarded to existing inbound
+liquidity if they can fit).
+
+`lsps2.holdincoming` takes parameters:
+
+```JSON
+{
+  "hold": true
+}
+```
+
+`hold` is an optional boolean.
+If unspecified, it defaults to `true`.
+If `true`, then all incoming payments will be held in PPPs
+even if they could fit in existing inbound liquidity.
+If `false`, this hold state is dropped, but existing PPPs
+remain in PPPs.
+
+`lsps2.holdincoming` returns no results `{}` and has no
+errors defined other than `unsupported_version`.
+
+> **Rationale** A better interface would have been to expose a
+> `parttoppp` call, where any HTLC or PTLC from LSP to client
+> that is failed by the client is converted to a PPP at the LSP
+> instead of propagating the error backwards, until the client
+> disables the `parttoppp` mode.
+> Such an interface would have lower payment latency.
+> Unfortunately, as of the time of this writing, no node software
+> has a hook or interface that allows capturing payment failures
+> and preventing their propagation back to the payer, so that
+> API would be difficult to implement without extensive changes
+> to existing node software.
+
+### 3. Pending Payment Part Notification
+
+If the client has previously called `lsps2.reservescid`, this also
+enables the LSP sending the notification `lsps2.youhaveppps`.
+The notification contains no parameters `{}`.
+
+This notification means that the client currently has pending payment
+parts (PPPs).
+
+The notification is sent depending on the state of the "PPP Queue"
+described in the next section.
+
+* On client reconnection, if the PPP Queue is currently non-empty.
+* When the PPP Queue transitions from empty to non-empty.
+
+When the client receives this notification, it should take items
+from the PPP Queue until it is emptied, persisting each item.
+
+> **Rationale** Instead of just a triggering notification that
+> contains no data, we could have designed this so that the PPP
+> details are already in the notification.
+>
+> The problem with that design is that due to the vagaries of TCP
+> sending, even if the LSP sends out a notification, it cannot
+> know if the client actually received the notification unless
+> the client actually reacts to that information.
+> If the LSP sees the client disconnect and then reconnect, either
+> it has to resend notifications for all pending PPPs (which might
+> be pointless in case the client already received it), or the
+> client and LSP need some kind of set-reconciliation algorithm
+> (which might end up being very complicated).
+>
+> Instead, the design in this specification uses an explicitly
+> queue, with the client doing explicit calls to inform the
+> LSP that it has received, processed, and persisted a PPP
+> before proceeding with the next one.
+
+### 4. Managing Pending Payment Parts
```

Comment:
I wanted to avoid replaying on reconnection, considering the possibility that the paye***r*** might have split the payment into a few hundred parts and the LSP might not have received all of them yet, then an interruption occurs on the client connection.  But maybe that is low-probability enough that it will not matter in practice?

Reply:
>It would need some handling for htlcs timeouts, especially for multipart payments.

What special handling would this need?  `failppp` would allow clients to do something like send a `mpp_timeout` failure.

Reply:
> I wanted to avoid replaying everything on reconnection, considering the possibility that the paye_**r**_ might have split the payment into a few hundred parts and the LSP might not have received all of them yet, then an interruption occurs on the client connection. But maybe that is low-probability enough that it will not matter in practice?

Reply:
> I wanted to avoid replaying everything on reconnection, considering the possibility that the paye_**r**_ might have split the payment into a few hundred parts and the LSP might not have received all of them yet, then an interruption occurs on the client connection. But maybe that is low-probability enough that it will not matter in practice?

I understand. Most ppps will be handled very quickly with a decision from the client. Basically all ppps need to get a response from the client what to do with it. We could add an intermediate decision for multipart payments, like an ACK, that the client has received it, if it doesn't know what to do with it just yet (if the LSP receives the ack that should mean the client persisted it and it doesn't have to replay that part)

Reply:
> What special handling would this need? `failppp` would allow clients to do something like send a `mpp_timeout` failure.

I can see how a client can request the LSP to forward the ppp, but when requested, the htlc timed out. In a normal lightning flow the client would not learn about this htlc at all if it timed out, but in this flow the client knows about it and it should be removed from the state machine somehow. Maybe a message from the LSP again: sorry this ppp you requested me to forward I already failed back.

Reply:
The queue is basically that ACK that the client has received it --- calling `popthenpeekppp` is basically "ACK the current PPP I have persisted it, gimme the next one".  If we want to ACK at all then I think a queue makes better sense.

I like not having a queue and just replaying everything on reconnect; if we replay everything that the client has not handled yet then the client does not have to persist anything at all, everything is just persisted at the LSP and the client can keep its copy of the PPPs in-memory, not in-disk, because if the client crashes, on reconnect the LSP just re-feeds the entire PPP set to the client.  In case it is the network that is the problem (i.e. client did not crash but mobile user entered a tunnel and lost carrier signal) then the re-send is idempotent, as the client should be able to look them up by PPP identifier and see it already has that PPP in-memory and can disregard the extra notification.

Reply:
>Maybe a message from the LSP again: sorry this ppp you requested me to forward I already failed back.

Yes, that is indeed what I plan to do.  Basically, instead of `retryppp` we have a `forwardppp` which accepts an array of PPP identifiers.  `retryppp` has no errors, but `forwardppp` has to have those.

LSP atomically:

* Checks that all identifiers are still in the PPP map.
  * If this check fails, error with `unknown_ppp` (2) with a `data.unknowns` field, an array containing the PPP identifiers that are in the params but not in the PPP map anymore --- basically this is the "I already failed them back".
* Checks that it can get capacity on the channel to the client for all the PPPs, and reserves those capacity for the PPPs.
  * If this check fails, error with `insufficient_funds` (3).  The client SHOULD then proceed to `getopeningfee` and `openwithppp`.
* Otherwise, disables the hold counter (so that the incoming HTLCs are now held) and arranges to forward the specified HTLCs to the client on channels with the client, and returns with a success.

So in this flow, the client always attempts `forwardppp` once it has completed a set of multipath PPPs (or once it gets a PPP that is a single-path payment).  If that fails with `unknown_ppp` it reupdates its in-memory map of PPPs to remove those PPPs that have been failed at the LSP.  If that fails with `insufficient_funds` it then proceeds to `getopeningfee` and (if it likes the opening feerate) `openwithppp`.

Reply:
> I like not having a queue and just replaying everything on reconnect

I agree

Reply:
> > Maybe a message from the LSP again: sorry this ppp you requested me to forward I already failed back.
> 
> Yes, that is indeed what I plan to do. Basically, instead of `retryppp` we have a `forwardppp` which accepts an array of PPP identifiers. `retryppp` has no errors, but `forwardppp` has to have those.
> 
> LSP atomically:
> 
> * Checks that all identifiers are still in the PPP map.
>   
>   * If this check fails, error with `unknown_ppp` (2) with a `data.unknowns` field, an array containing the PPP identifiers that are in the params but not in the PPP map anymore --- basically this is the "I already failed them back".
> * Checks that it can get capacity on the channel to the client for all the PPPs, and reserves those capacity for the PPPs.
>   
>   * If this check fails, error with `insufficient_funds` (3).  The client SHOULD then proceed to `getopeningfee` and `openwithppp`.
> * Otherwise, disables the hold counter (so that the incoming HTLCs are now held) and arranges to forward the specified HTLCs to the client on channels with the client, and returns with a success.
> 
> So in this flow, the client always attempts `forwardppp` once it has completed a set of multipath PPPs (or once it gets a PPP that is a single-path payment). If that fails with `unknown_ppp` it reupdates its in-memory map of PPPs to remove those PPPs that have been failed at the LSP. If that fails with `insufficient_funds` it then proceeds to `getopeningfee` and (if it likes the opening feerate) `openwithppp`.

This sounds very good to me.

I wonder whether the part `and reserves those capacity for the PPPs` is doable today or whether we'd need to modify lightning implementations for this. 

What do you mean with `disables the hold counter (so that the incoming HTLCs are now held)`? Do you mean it resumes those htlcs now?

Reply:
>I wonder whether the part `and reserves those capacity for the PPPs` is doable today or whether we'd need to modify lightning implementations for this.

If the LSP disallows use of the "real" SCID of the channels to the client, and only allows use of the issued Continuous JIT permanent/ephemeral SCIDs, then the LSP code should be able to keep track of how much of the capacity is in use in the channel by the Continuous JIT system.  Then it simply needs a single counter, protected by some kind of atomicity (a single global lock for all Continuous JIT channels with the client) to reserve this capacity, without having to modify the internals of the Lightning implementation.  This is because only the Continuous JIT system is the one that actually forwards to the clients over those channels, so it should be able to accurately account for every millisatoshi of capacity.

This should be doable at least on CLN --- `htlc_received` hook would let you filter out forwards that go to the "real" SCID, and if you detect an issued permanent/ephemeral SCID, you can then keep track of how much capacity you have in your channel with the client that is not yet reserved for HTLCs.  There is the bit of complication in that actual HTLCs have to pay onchain fees (which means that the amount you have to reserve is "HTLC amount + HTLC onchain fee" --- and you might need to override your onchain fee-reporting in the CLN interfface to the blockchain to at least temporarily stop updating while you are calculating this).  You can even ensure atomicity of the amount being reserved (in case of multiple parallel running `forwardppps` from the client) by using the CLN interface to the database, which includes an incrementing version number you can use to ensure that your data (in this context the capacity of the channel(s) with the client) is not stale before you overwrite it with new data.

I suspect that LND would be broadly similar, and it should be easier with LDK.

> What do you mean with `disables the hold counter (so that the incoming HTLCs are now held)`? Do you mean it resumes those htlcs now?

We write elsewhere that LSPs should hold inbound HTLCs to be forwarded to permanent/ephemeral SCIDs, for 60 to 120 seconds at most.  That is the timer that should be disabled at this point.

Reply:
Against this, we should note that the LSP might reserve some outbound liquidity on one or more channels to a set of HTLC, but adding multiple HTLCs is not a single atomic operation, especially if multiple channels are involved.  The following is possible:

* Client+LSP have 2 channels with inbound liquidity.
* Client asks two PPPs to be forwarded in a single `forwardppp` call, intended to be atomic.
* LSP schedules one PPP on one channel, the other PPP on the other channel.
* Client signs off (`commitment_signed`+`revoke_and_ack`) on the new PPP on one channel.
* Before the client can sign off on the other channel it gets closed --- maybe client is messing around, or LSP has automated system (separate from the PPP-handling component) that periodically considers channels for closure.  Thus the channel is closed but with the other PPP not instantaited as an HTLC on that channel.

So we specify LSP SHOULD best-effort send all the HTLCs or PTLCs together, but if is unable to (due to the above race condition), it SHOULD return the H/PTLC(s) that did not make it through to the PPP Map, with new PPP identifiers and with the hold timer continued, and with the notification re-sent.  The client MUST check for this condition occurring.